worker/interactions.go

```diff
@@ -0,0 +1,247 @@
+	for _, m := range mr.ms {
+		his = append(his, metricToInteraction(m))
+	}
+	return his
+}
+
+// metricCommon contains the common fields of all metrics.
+type metricCommon struct {
+	hostKey   types.PublicKey
+	address   string
+	timestamp time.Time
+	elapsed   time.Duration
+	err       error
+}
+
+type metricResultCommon struct {
+	Address   string        `json:"address"`
+	Timestamp time.Time     `json:"timestamp"`
+	Elapsed   time.Duration `json:"elapsed"`
+}
+
+func (m metricCommon) commonResult() metricResultCommon {
+	return metricResultCommon{
+		Address:   m.address,
+		Timestamp: m.timestamp,
+		Elapsed:   m.elapsed,
+	}
+}
+func (m metricCommon) HostKey() types.PublicKey { return m.hostKey }
+func (m metricCommon) Timestamp() time.Time     { return m.timestamp }
+
+func (m metricCommon) IsSuccess() bool {
+	return isSuccessfulInteraction(m.err)
+}
+
+// MetricHostDial is a metric that contains the result of a dial attempt.
+type MetricHostDial struct {
+	metricCommon
+}
+
+func (m MetricHostDial) Result() interface{} {
+	cr := m.commonResult()
+	er := hostdb.ErrorResult{Error: errToStr(m.err)}
+	if m.err != nil {
+		return struct {
+			metricResultCommon
+			hostdb.ErrorResult
+		}{cr, er}
+	} else {
+		return struct {
+			metricResultCommon
+		}{cr}
+	}
+}
+
+func (m MetricHostDial) Type() string { return "dial" }
+
+// MetricPriceTableUpdate is a metric that contains the result of fetching a
+// price table.
+type MetricPriceTableUpdate struct {
+	metricCommon
+	pt hostdb.HostPriceTable
+}
+
+func (m MetricPriceTableUpdate) Result() interface{} {
+	cr := m.commonResult()
+	er := hostdb.ErrorResult{Error: errToStr(m.err)}
+	if m.err != nil {
+		return struct {
+			metricResultCommon
+			hostdb.ErrorResult
+		}{cr, er}
+	} else {
+		return struct {
+			metricResultCommon
+			hostdb.PriceTableUpdateResult
+		}{cr, hostdb.PriceTableUpdateResult{ErrorResult: er, PriceTable: m.pt}}
+	}
+}
+
+func (m MetricPriceTableUpdate) Type() string {
+	return hostdb.InteractionTypePriceTableUpdate
+}
+
+// MetricHostScan is a metric that contains the result of a host scan.
+type MetricHostScan struct {
+	metricCommon
+	pt       rhpv3.HostPriceTable
+	settings rhpv2.HostSettings
+}
+
+func (m MetricHostScan) Result() interface{} {
+	cr := m.commonResult()
+	er := hostdb.ErrorResult{Error: errToStr(m.err)}
+	if m.err != nil {
+		return struct {
+			metricResultCommon
+			hostdb.ErrorResult
+		}{cr, er}
+	} else {
+		return struct {
+			metricResultCommon
+			hostdb.ScanResult
+		}{cr, hostdb.ScanResult{ErrorResult: er, PriceTable: m.pt, Settings: m.settings}}
+	}
+}
+
+func (m MetricHostScan) Type() string {
+	return hostdb.InteractionTypeScan
+}
+
+func isSuccessfulInteraction(err error) bool {
+	// No error always means success.
+	if err == nil {
+		return true
+	}
+	// List of errors that are considered successful interactions.
+	if errors.Is(err, ErrInsufficientFunds) || strings.Contains(err.Error(), ErrInsufficientFunds.Error()) {
```

Comment:
did you forget to remove this? you have `isBalanceInsufficient`

Reply:
`ErrInsufficientFunds` is returned by `payByContract` when the contract doesn't have enough funds.
The other one is for EA balances.